cloudformation/memsub-promotions-lambdas-cf.yaml

```diff
@@ -0,0 +1,253 @@
+            Action:
+            - dynamodb:BatchWriteItem
+            - dynamodb:PutItem
+            Resource:
+            - arn:aws:dynamodb:*:*:table/MembershipSub-PromoCode-View-PROD
+            - arn:aws:dynamodb:*:*:table/MembershipSub-PromoCode-View-UAT
+            - arn:aws:dynamodb:*:*:table/MembershipSub-PromoCode-View-DEV
+  MembershipSubPromotionsDataRole:
+    Type: AWS::IAM::Role
+    Properties:
+      AssumeRolePolicyDocument:
+        Statement:
+        - Effect: Allow
+          Principal:
+            Service:
+            - lambda.amazonaws.com
+          Action:
+          - sts:AssumeRole
+      Path: /
+      Policies:
+      - PolicyName: MembershipSub-Promotions-Data-Policy
+        PolicyDocument:
+          Version: '2012-10-17'
+          Statement:
+          - Effect: Allow
+            Action:
+            - lambda:InvokeFunction
+            - logs:CreateLogGroup
+            - logs:CreateLogStream
+            - logs:PutLogEvents
+            Resource: '*'
+          - Effect: Allow
+            Action:
+            - dynamodb:GetRecords
+            - dynamodb:Scan
+            Resource:
+            - arn:aws:dynamodb:*:*:table/MembershipSub-PromoCode-View-PROD
+            - arn:aws:dynamodb:*:*:table/MembershipSub-PromoCode-View-UAT
+            - arn:aws:dynamodb:*:*:table/MembershipSub-PromoCode-View-DEV
+          - Effect: Allow
+            Action:
+            - s3:PutObject
+            - s3:PutObjectAcl
+            Resource:
+            - arn:aws:s3:::ophan-raw-membership-promo-code-view/*
+  PromoCodeViewLambdaFunctionDEV:
+    Type: AWS::Lambda::Function
+    Properties:
+      Code:
+        S3Bucket: gu-promotions-tool-dist
+        S3Key: membership/PROD/MembershipSub-Promotions-PromoCode-View/MembershipSub-Promotions-PromoCode-View.zip
+      Description: An Amazon DynamoDB trigger that creates a view of data associated
+        with a particular promo code for the data pipeline to collect.
+      FunctionName: MembershipSub-Promotions-to-PromoCode-View-DEV
+      Handler: MembershipSub-Promotions-to-PromoCode-View-Lambda.handler
+      MemorySize: 128
+      Role: !GetAtt [MembershipSubPromotionsToolRole, Arn]
+      Runtime: nodejs6.10
+      Timeout: 10
+  PromoCodeViewLambdaFunctionUAT:
+    Type: AWS::Lambda::Function
+    Properties:
+      Code:
+        S3Bucket: gu-promotions-tool-dist
+        S3Key: membership/PROD/MembershipSub-Promotions-PromoCode-View/MembershipSub-Promotions-PromoCode-View.zip
+      Description: An Amazon DynamoDB trigger that creates a view of data associated
+        with a particular promo code for the data pipeline to collect.
+      FunctionName: MembershipSub-Promotions-to-PromoCode-View-UAT
+      Handler: MembershipSub-Promotions-to-PromoCode-View-Lambda.handler
+      MemorySize: 128
+      Role: !GetAtt [MembershipSubPromotionsToolRole, Arn]
+      Runtime: nodejs6.10
+      Timeout: 10
+  PromoCodeViewLambdaFunctionPROD:
```

Comment:
I find it odd that we have DEV/CODE/PROD lambdas all in the same CFN - wouldn't it be better to have a stack for each and use Stage as a parameter?

Reply:
I agree in principle, however we don't create CFN stages (PROD/CODE) directly as per Touchpoint backends which is what these Lambdas align to. This separation out into separate files goes at least gets rid of the previous Conditionals.

Reply:
The code for all 3 lambda are updated via a PROD deploy of RiffRaff. I could tweak the Lambda's S3 bucket locations so that the DEV lambda reads the DEV S3 RiffRaff artefact and UAT the UAT S3 artefact? However given the code is in node, the DEV lambdas can be developed in the console or even cloned and developed. I'll leave this till next time though as it's not the concern of this change.

Reply:
Sounds good, thanks for explaining this 👍